src/officehours_api/backends/backend_phaser.py

```diff
@@ -0,0 +1,67 @@
+import logging
+from typing import List
+
+from django.conf import settings
+from django.db.models import QuerySet
+
+from officehours_api.models import Meeting, MeetingStatus, Queue
+
+
+logger = logging.getLogger(__name__)
+
+
+class BackendPhaser:
+    """
+    Class for managing queries and operations related to the phasing out of backends.
+    """
+
+    def __init__(self, old_backend: settings.IMPLEMENTED_BACKEND):
+        self.old_backend: settings.IMPLEMENTED_BACKEND = old_backend
+
+    def get_queues_allowing_backend(self) -> QuerySet:
+        return Queue.objects.filter(allowed_backends__contains=[self.old_backend])
+
+    def get_meetings_with_backend(self) -> QuerySet:
+        return Meeting.objects.filter(backend_type=self.old_backend)
+
+    def remove_backend_from_queue_allowed_backends(self):
+        queues_allowing_backend = self.get_queues_allowing_backend()
+        logger.info(f'Number of queues allowing {self.old_backend}: {len(queues_allowing_backend)}')
+        for queue in queues_allowing_backend:
+            queue.remove_allowed_backend(self.old_backend)
+            logger.info(f'Removed {self.old_backend} as an allowed backend from queue {queue.id}')
+
+    def set_unstarted_meetings_to_default_backend(self):
+        logger.info('Setting backend_type of unstarted meetings to use the default backend...')
+        meetings_with_backend = self.get_meetings_with_backend()
+        unstarted_meetings_with_old_backend: List[Meeting] = [
+            meeting for meeting in meetings_with_backend if meeting.status != MeetingStatus.STARTED
+        ]
+        for meeting in unstarted_meetings_with_old_backend:
+            meeting.change_backend_type()
+        logger.info(
+            f'Set the backend_type for {len(unstarted_meetings_with_old_backend)} meeting(s) '
+            f'to the default {settings.DEFAULT_BACKEND}.'
+        )
```

Comment:
This is all good, but an errant thought that popped into my head is that when I write code like this, I generally think logging is better suited to calling contexts (like the view layer or the CLI layer, or even just hoisting it into an aggregate function like `phase_out`), while deeper contexts are better off returning information about the work they did (for example, in this case you could return `unstarted_meetings_with_old_backend` wholesale, then the calling context can get its `len` and compose the log message). This offers better control over logging and makes your functions more functional, which in turn makes them more flexible.

That said, it wouldn't provide any major immediate benefit in this situation (we're unlikely to reuse this code elsewhere), but since it was on my mind I thought I'd offer it up as food for thought.

Reply:
So in summary, you're saying return the data needed for logging and then log in `phase_out` or at the CLI level?

In terms of knowing exactly what the program is doing, I see some benefit to having some logging at the point of doing the work. You can have log messages tell you where you it's being called from (the method), so people can quickly find the relevant code if there's a problem. All of these could get downgraded to `debug`.

I think I follow your reasoning regarding having logging at a higher level. However, do you see any downside to separating the acting logic and the log creation? I wonder if returning fairly different data from all these functions to a higher-order method/function may make that code more difficult to parse for humans. I have seen people pass strings back and either log that or concatenate it with other strings to create a report at the end; I've never really loved that either.

I appreciate you raising this, it's good to think about these patterns more.

Reply:
> So in summary, you're saying return the data needed for logging and then log in phase_out or at the CLI level?

Maybe; I realize this isn't a clear example of where it would be helpful. If you picture a method that looks at a data structure and conditionally changes some of its elements, you could imagine a lot of situations where calling contexts might want to do something other than just log the ones that were changed, or log them differently. So returning the changed records could potentially be more powerful, or returning a new data structure leaving the old one intact (immutability)... 

This idea also kinda relates to a concept in FP where you want to avoid "side effects" (like logging, network calls, etc.) but those "side effects" actually compose all of the useful stuff your program does (writing stuff to the screen, writing an http response...), so what you really do is hoist those parts up to the top layer of your app, making the lower layers more data driven.

> I see some benefit to having some logging at the point of doing the work

Agree, although if your functions are small and simple and pure enough, unit testing or even just REPL testing becomes a lot easier.

> I wonder if returning fairly different data from all these functions to a higher-order method/function may make that code more difficult to parse for humans

Yeah, this is particularly a problem in Python where typings aren't very good yet (being able to infer method types in your IDE helps immensely with following the shape of the data), so I agree that it's not a silver bullet, just something to think about.

Reply:
Okay, does something like this seem better to you? https://github.com/tl-its-umich-edu/remote-office-hours-queue/pull/305/commits/fa24545b6ce21857a9163e2553dffe67aa0fba3f

Reply:
Maybe ever so slightly; you didn't need to change anything, but LGTM!